### PR TITLE
Python: Use `TUnknown` as the result of calls to methods with unknown return types

### PR DESCRIPTION
**Pull request by tausbn**
TL;DR: We were not handling calls to methods like `dict.get` correctly, and thus they would not receive a corresponding `Value`. The change is fairly simple: If we don't know the precise return type, just assume it's `object`. The pointed-to `Value` will then be an unknown instance of `object`, which is fine.

### CODE REVIEW

File: python/ql/src/semmle/python/objects/Callables.qll [hunk 1]
```diff
@@ -333,10 +333,19 @@ class BuiltinMethodObjectInternal extends CallableObjectInternal, TBuiltinMethod
...
+        /* If we have a record of the return type in our stubs, use that. */
         exists(Builtin func |
             func = this.getBuiltin() |
             ext_rettype(func, result)
         )
+        or
+        /* Otherwise, if no such record exists, use `object` as the return type. */
+        not exists(Builtin func |
+            // We cannot do `this.getBuiltin()` here, as that would introduce negative recursion.
+            // Instead, we appeal directly to the underlying IPA type.
```
Comment (RasmusWL) [inline python/ql/src/semmle/python/objects/Callables.qll]: Nitpick: was is up with the mix of multiline and singleline comments? :D

File: python/ql/test/library-tests/PointsTo/new/PointsToMissing.ql [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import python
+import Util
+import semmle.python.pointsto.PointsTo
+import semmle.python.objects.ObjectInternal
+
+/* Ideally, this test should return _no_ results. */
```
Comment (RasmusWL) [inline python/ql/test/library-tests/PointsTo/new/PointsToMissing.ql]:
```suggestion
/* This test should return _no_ results. */
```

When I read the _Ideally_ part, I assumed that it actually would return results :confused:

File: python/ql/src/semmle/python/objects/Callables.qll (hunk 1 above)
Comment (tausbn) [reply python/ql/src/semmle/python/objects/Callables.qll]: Hey, man, I'm a free spirit. You can't tie me down to a single style of comments!

But seriously, there's no reasoning behind it. I think my preference is the `/* ... */` style, but of course I couldn't use that on lines 344-345 as autoformat would expand that into four lines. So in an effort to be forward-thinking, I used line comments there instead.

**Comment by RasmusWL** [comment]
I'm actually curious if this would have any performance implications. I don't have an intuitive feel for it, but I'm guessing that anything that changes points-to analysis might be worth investigating? :smile:

**Comment by tausbn** [comment]
You're right. I was considering this myself. I'll set up a differences job to see how it fares.

**Comment by tausbn** [comment]
Dist-compare report: https://git.semmle.com/gist/taus/269cda979d7c3d34ebbbf12d4f860eca
_Lots_ of changes in alerts (probably because there is way less points-to pruning now). I still haven't gone through these results to see if they're correct.

**Comment by tausbn** [comment]
Okay, having now looked at the results, I see some problems with them. Currently, we do not consider an unknown instance of `object` to be callable, so we get lots of false positives for non-callable called.
I'll see if `TUnknown` fares better in this aspect.

**Comment by tausbn** [comment]
Note: I expect tests to start failing again, since a bunch of `Value`s will now be missing (as we do not at present allow unknown values). I will fix this up in a later commit, as well as address the review comments.

File: python/ql/test/library-tests/PointsTo/new/PointsToMissing.ql (hunk 2 above)
Comment (tausbn) [reply python/ql/test/library-tests/PointsTo/new/PointsToMissing.ql]: Done!

**Comment by RasmusWL** [comment]
I think it looks good, but could we do an other dist-compare after the ```Python: Use `TUnknown` instead of `TUnknownInstance`.``` commit?

**Comment by tausbn** [comment]
Indeed. That's why I left the "Awaiting Evaluation" label. Hopefully we should have fresh results tomorrow. :slightly_smiling_face:

**Comment by tausbn** [comment]
Latest report: https://git.semmle.com/gist/taus/07875fdb07c028c31a0a7c06a26eab44
Haven't looked at it in detail yet, but it looks decent enough.

**Comment by RasmusWL** [comment]
Except for that merge conflict of course :smile: (probably just got introduced since I just merged 2 PRs)

**Comment by RasmusWL** [comment]
Oh no, tests are failing :cry: @tausbn

**Pull request merged**